hedera-node/src/main/java/com/hedera/services/store/contracts/HederaStackedWorldStateUpdater.java

```diff
@@ -180,6 +219,19 @@ private boolean isMissingTarget(final Address alias) {
 		return !trackingAccounts().exists(accountIdFromEvmAddress(target));
 	}
 
+	private HederaWorldState getHederaWorldState() {
+		var parentUpdater = parentUpdater();
+		while (parentUpdater.isPresent() && parentUpdater.get().parentUpdater().isPresent()) {
+			parentUpdater = parentUpdater.get().parentUpdater();
+		}
+		if (parentUpdater.isEmpty()) {
+			throw new InvalidTransactionException(ResponseCodeEnum.FAIL_INVALID);
+		}
+
+		final var topLevelUpdater = (HederaWorldState.Updater) parentUpdater.get();
+		return (HederaWorldState) topLevelUpdater.wrappedWorldView();
+	}
+
```

Comment:
Is this the best approach we can take?

Another would be to extract both `WorldStateAccount` and `WorldStateTokenAccount` from `HederaWorldState` into their own separate classes, which would be used both here and in `HederaWorldState`.

Reply:
I think it's better to extract `WorldStateTokenAccount` in its own separate class that has **no** access to state. `WorldStateAccount` should probably remain in `HederaWorldState` since it tightly-coupled to the state held there. My reasoning: 
- we only need to know that the address we're loading is a token address and we should have that info in the current or parent updater, no need to go to the top-level one
- we use the `WorldStateTokenAccount` type to hardcode the redirect bytecode returned by the `getCode()` method, this is a constant one that doesn't need to load from state
- we never want to give access to the world state of a specific updater (especially the top-level one) in a child updater, and creating `WorldStateTokenAccount` object from the context of `HederaWorldState` may leave a backdoor open 
- it would be simpler and easier to read/understand in my opinion

Is there a specific reason not to extract `WorldStateTokenAccount` in it's own class?

Reply:
Yup, extracting `WorldStateTokenAccount` into a separate class was the initial plan. However, since the [HederaWorldState get() method](https://github.com/hashgraph/hedera-services/blob/f0c1a822b4ced40155074bfd5dd2933ca7dd65ec/hedera-node/src/main/java/com/hedera/services/store/contracts/HederaWorldState.java#L178-L180) (which up until this point has been used for redirecting) returns a `WorldStateAccount`, the newly extracted `WorldStateTokenAccount` has to extend `WorldStateAccount`, but since the latter is an inner class that is not possible 😞

Reply:
Oh, I see... then what about creating new special subtypes of `Account` and `EvmAccount` which encapsulate the hardcoded bytecode + the token address and maybe throw exceptions on any state change attempt?

Reply:
Hmm, just to make sure we are on the same page --- the idea is to create a new class/classes that will be used **solely** in the stacked updaters and also leave `WorldStateTokenAccount`, right?

Reply:
What about making WorldStateTokenAccount implements Account, instead of extending MerkleStateAccount. In this case it would be possible to extract WorldStateTokenAccount in a separate class.

Reply:
In this way we would also be able to extract TOKEN_BYTECODE_PATTERN and TOKEN_CALL_REDIRECT_CONTRACT_BINARY in this new class.

Reply:
Wouldn't that be impossible again, though, since in this case `WorldStateTokenAccount` wouldn't be extending `WorldStateAccount` and we wouldn't be able to return it from the `get()` in `HederaWorldState` (which returns a `WorldStateAccount`) or am I missing something?

Reply:
We can explore the option `HederaWorldState.get()` to return `Account` interface. If we get one step back, we can see that `AbstractLedgerWorldUpdate.get()` already returns `Account` interface.

Reply:
I was thinking:
- not changing the behaviour in `HederaWorldState`
- adding new token "stubs" - probably an implementation of `EvmAccount` with implementation for `Bytes getCode()` returning the redirect bytecode and all other methods throwing an exception (we don't expect to call them on a token address AFAICT)
- returning the stub from `AbstractLedgerWorldUpdater` `get(address)` and `getAccount(address)`

Reply:
That will work, with the expense of introducing some amount of duplication/redundancy... I agree that changing `HederaWorldState` should be treated very carefully and if we were to do it it would require more examination and research  on breaking existing behavior

Reply:
@tinker-michaelj can you advise what you think the best approach to solve this would be in your opinion?
TLDR: is the solution proposed by @dimitar-dinev acceptable, or should we go for a "stubs" approach?